### PR TITLE
feat: [PL-43172]: Create resource api's for notification rule and cha…

### PR DESCRIPTION
**Pull request by jennyjames21**
Summary:
-Resource APIs for Notification Rule and Notification Channel

Test
- API test for create, update, get, search and delete notification channel 
- API test for create, update, get, search and delete notification rule 

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56141)

### CODE REVIEW

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> Reviewed 9 of 22 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56141)* status: 22 file(s) to review. 1 change requests. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationChannelManagementServiceImpl.java` line 81 at r2](https://reviewable.io/reviews/harness/harness-core/56141#-NkZ35Wj78wavTXbWIMY:-NkZ35Wj78wavTXbWIMZ:b-3ltjyq) ([raw file](https://github.com/harness/harness-core/blob/a1e9f0db98c014195304c96487b537425c58caa9/platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationChannelManagementServiceImpl.java#L81)):*
> ```Java
> 
>   @Override
>   public void checkUniqueness(String accountIdentifier, String orgIdentifier, String projectIdentifier,
> ```

Should be private
___
*[`platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationRuleManagementServiceImpl.java` line 93 at r2](https://reviewable.io/reviews/harness/harness-core/56141#-NkZ30kIEs4oSS2vLvHq:-NkZ30kJ7kLYECI4f507:b-3ltjyq) ([raw file](https://github.com/harness/harness-core/blob/a1e9f0db98c014195304c96487b537425c58caa9/platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationRuleManagementServiceImpl.java#L93)):*
> ```Java
> 
>   @Override
>   public void checkUniqueness(
> ```

Should be private
___
*[`platform-service/modules/notification-service/src/main/java/io/harness/notification/utils/NotificationManagementServiceUtils.java` line 26 at r2](https://reviewable.io/reviews/harness/harness-core/56141#-NkZ3dbq79S6hwWpF8if:-NkZ3dbq79S6hwWpF8ig:b-k2ejfw) ([raw file](https://github.com/harness/harness-core/blob/a1e9f0db98c014195304c96487b537425c58caa9/platform-service/modules/notification-service/src/main/java/io/harness/notification/utils/NotificationManagementServiceUtils.java#L26)):*
> ```Java
> import org.springframework.data.domain.Pageable;
> 
> public class NotificationManagementServiceUtils {
> ```

Rename to NotificationManagementApiUtils if it only has methods required for API resource layer.
___
*[`platform-service/modules/notification-service/src/main/java/io/harness/notification/utils/NotificationManagementServiceUtils.java` line 63 at r2](https://reviewable.io/reviews/harness/harness-core/56141#-NkZ4-Jf76_UEPUbs26S:-NkZ4-JgEqzuoML3OzAL:b-ihw643) ([raw file](https://github.com/harness/harness-core/blob/a1e9f0db98c014195304c96487b537425c58caa9/platform-service/modules/notification-service/src/main/java/io/harness/notification/utils/NotificationManagementServiceUtils.java#L63)):*
> ```Java
>     }
>     switch (sortField) {
>       case IDENTIFIER:
> ```

Just check if this is required or not.  Check required mongo index is present.

**Comment by jennyjames21** [comment]
*[`platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationChannelManagementServiceImpl.java` line 81 at r2](https://reviewable.io/reviews/harness/harness-core/56141#-NkZ35Wj78wavTXbWIMY:-NkZBRtREBt6Alk7UUGG:bvnrpcm) ([raw file](https://github.com/harness/harness-core/blob/a1e9f0db98c014195304c96487b537425c58caa9/platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationChannelManagementServiceImpl.java#L81)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Should be private
</blockquote></details>

updated

**Comment by jennyjames21** [comment]
*[`platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationRuleManagementServiceImpl.java` line 93 at r2](https://reviewable.io/reviews/harness/harness-core/56141#-NkZ30kIEs4oSS2vLvHq:-NkZBV2nBnYKyQmFZO9G:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/a1e9f0db98c014195304c96487b537425c58caa9/platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationRuleManagementServiceImpl.java#L93)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Should be private
</blockquote></details>

done

**Comment by jennyjames21** [comment]
*[`platform-service/modules/notification-service/src/main/java/io/harness/notification/utils/NotificationManagementServiceUtils.java` line 63 at r2](https://reviewable.io/reviews/harness/harness-core/56141#-NkZ4-Jf76_UEPUbs26S:-NkZBYtO9nMAntpwdC0i:bvogskk) ([raw file](https://github.com/harness/harness-core/blob/a1e9f0db98c014195304c96487b537425c58caa9/platform-service/modules/notification-service/src/main/java/io/harness/notification/utils/NotificationManagementServiceUtils.java#L63)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Just check if this is required or not.  Check required mongo index is present.
</blockquote></details>

removing for now, will add if needed

File: platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml [hunk 1]
```diff
@@ -58,14 +58,19 @@ paths:
...
       parameters:
         - $ref: '#/components/parameters/AccountHeader'
         - $ref: '#/components/parameters/Page'
         - $ref: '#/components/parameters/Limit'
         - $ref: '#/components/parameters/Sort'
         - $ref: '#/components/parameters/Order'
+        - schema:
+            type: string
+          in: query
+          name: search_term
```
Comment (mankrit-singh) [inline platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:72]: could you use the globally defined search term?

**Review by mankrit-singh — APPROVED** [review]

File: platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml (hunk 1 above)
Comment (jennyjames21) [reply platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml]: updated

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger uts

**Comment by jennyjames21** [comment]
trigger ut9

**Comment by jennyjames21** [comment]
trigger messagecheck

**Review by mankrit-singh — APPROVED** [review]

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> Reviewed 5 of 22 files at r1, 7 of 8 files at r3, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56141)* status: 22 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`platform-service/modules/notification-service/src/main/java/io/harness/notification/remote/resources/v1/NotificationRuleApiImpl.java` line 55 at r6](https://reviewable.io/reviews/harness/harness-core/56141#-NknAMtABZwOcl1xJnr4:-NknAMtABZwOcl1xJnr5:boflocu) ([raw file](https://github.com/harness/harness-core/blob/2f1ec482e697db3522fd377a5c4b481a7e38ccdc/platform-service/modules/notification-service/src/main/java/io/harness/notification/remote/resources/v1/NotificationRuleApiImpl.java#L55)):*
> ```Java
>     NotificationRule notificationRule =
>         notificationRuleManagementService.create(notificationServiceManagementMapper.toNotificationRuleEntity(body));
>     return Response.status(Response.Status.CREATED).entity(notificationRule).build();
> ```

Why is the API returning entity object instead of DTO? Entity classes should be private to a service.
___
*[`platform-service/modules/notification-service/src/main/java/io/harness/notification/remote/resources/v1/NotificationRuleApiImpl.java` line 65 at r6](https://reviewable.io/reviews/harness/harness-core/56141#-Nkn9189Bh6zmua245-_:-Nkn9189Bh6zmua245-a:boflocu) ([raw file](https://github.com/harness/harness-core/blob/2f1ec482e697db3522fd377a5c4b481a7e38ccdc/platform-service/modules/notification-service/src/main/java/io/harness/notification/remote/resources/v1/NotificationRuleApiImpl.java#L65)):*
> ```Java
>     NotificationRule notificationRuleEntity =
>         notificationRuleManagementService.get(harnessAccount, org, project, notificationRuleIdentifier);
>     return Response.status(Response.Status.OK).entity(notificationRuleEntity).build();
> ```

Why is the API returning entity object instead of DTO? Entity classes should be private to a service.
___
*[`platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationChannelManagementServiceImpl.java` line 39 at r6](https://reviewable.io/reviews/harness/harness-core/56141#-NknA5yt9uPWTZ5LUdj0:-NknA5yt9uPWTZ5LUdj1:bo6a6o8) ([raw file](https://github.com/harness/harness-core/blob/2f1ec482e697db3522fd377a5c4b481a7e38ccdc/platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationChannelManagementServiceImpl.java#L39)):*
> ```Java
>   @Override
>   public NotificationChannel create(NotificationChannel notificationChannel) {
>     checkUniqueness(notificationChannel.getAccountIdentifier(), notificationChannel.getOrgIdentifier(),
> ```

We can avoid extra DB call in create flow by catching DuplicateKeyException as it was done earlier.
___
*[`platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationRuleManagementServiceImpl.java` line 41 at r6](https://reviewable.io/reviews/harness/harness-core/56141#-Nkn9sPg3_wsL94XN63S:-Nkn9sPhDDzFcNi-zizA:bo6a6o8) ([raw file](https://github.com/harness/harness-core/blob/2f1ec482e697db3522fd377a5c4b481a7e38ccdc/platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationRuleManagementServiceImpl.java#L41)):*
> ```Java
>   @Override
>   public NotificationRule create(NotificationRule notificationRule) {
>     checkUniqueness(notificationRule.getAccountIdentifier(), notificationRule.getOrgIdentifier(),
> ```

We can avoid extra DB call in create flow by catching DuplicateKeyException as it was done earlier.

**Comment by jennyjames21** [comment]
*[`platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationChannelManagementServiceImpl.java` line 39 at r6](https://reviewable.io/reviews/harness/harness-core/56141#-NknA5yt9uPWTZ5LUdj0:-NknGGyXAFe-CU43SRDV:bp9x9q) ([raw file](https://github.com/harness/harness-core/blob/2f1ec482e697db3522fd377a5c4b481a7e38ccdc/platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationChannelManagementServiceImpl.java#L39)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

We can avoid extra DB call in create flow by catching DuplicateKeyException as it was done earlier.
</blockquote></details>

uniqueness check is for identifier.(Account + org + prog  + indetifier). Note create notification is not frequent call, one time while creating notification rule in the account

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56141)* status: 22 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationChannelManagementServiceImpl.java` line 39 at r6](https://reviewable.io/reviews/harness/harness-core/56141#-NknA5yt9uPWTZ5LUdj0:-NknHKfW3JnXRhaZTs_L:b-zb4i1f) ([raw file](https://github.com/harness/harness-core/blob/2f1ec482e697db3522fd377a5c4b481a7e38ccdc/platform-service/modules/notification-service/src/main/java/io/harness/notification/service/NotificationChannelManagementServiceImpl.java#L39)):*
<details><summary><i>Previously, jennyjames21 (Jenny James) wrote…</i></summary><blockquote>

uniqueness check is for identifier.(Account + org + prog  + indetifier). Note create notification is not frequent call, one time while creating notification rule in the account
</blockquote></details>

Agree that its not a frequent call. But are we not defining a unique key for the the collection?

**Comment by jennyjames21** [comment]
trigger ut4

**Comment by jennyjames21** [comment]
trigger ut9

**Review by jennyjames21 — COMMENTED** [review]
> Reviewed 12 of 23 files at r1, 4 of 9 files at r3, 1 of 1 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, 5 of 5 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56141)* status: 24 file(s) to review. 1. unresolved discussions. Waiting on team(s): del-team.

**Review by bharatgoelharness — APPROVED** [review]

**Pull request merged**